docs/content/latest/develop/best-practices-ysql.md

```diff
@@ -0,0 +1,137 @@
+from Amazon AWS with added functionality of programmable Query Routing & Rewriting.
+4. Custom logic in your code
+You can keep a list of servers on the client, have a connection pool for each client and use them
+randomly or by a defined logic. You can consult our [JDBC driver](https://github.com/yugabyte/jdbc-yugabytedb) 
+for implementation details.
+5. [odyssey pooler](https://github.com/yandex/odyssey) as connection pooler/proxy.
+See [comparisons](https://github.com/yandex/odyssey/issues/3) to pgbouncer.
+6. [consul](https://www.consul.io/) can be used for service discovery of the tserver nodes
+On the client node the consul agent will provide a hostname which will point to a 
+server that is up. You can use that hostname in the connection string to YugabyteDB.
+This way you keep only 1 connection pool to 1 server. When the server goes down,
+consul will return a new hostname which will be able to serve queries.
+7. [nginx](http://nginx.org/) can be used as a tcp proxy & load balancer in front
+of a cluster of YugabyteDB.
+8. When using kubernetes, YugabyteDB uses it's load balancer to proxy the tservers.
+9. In GCP you can use [Cloud Load Balancer](https://cloud.google.com/load-balancing/) as TCP proxy.
+10. In AZURE you can use [Load Balancer](https://azure.microsoft.com/en-us/services/load-balancer/) service. 
+
+{{< note title="Note" >}}
+- Make sure to set forwarding rules to port :5433.
+- Configure health checks using HTTP port :7000 and path `/status`.
+- The load balancer is only needed for tservers. Clients don't need to connect to Masters.
+{{< /note >}}
+
+## Partial indexes
+Creating indexes actually creates another hidden table where 1 row is inserted for each row from the base table.
+This includes a lot of overhead when we're only interested on querying a subset of the values. 
+
+A common scenario is to query the index only when the value of a column `is not NULL`. 
+This can be achieved with partial indexes using the `WITH` clause when creating an index.
+
+The partial index has lower storage overhead and better write performance because not all inserts will also write to the index table.
+
+Partial indexes can also be combined with [Covering indexes](../api/ysql/commands/ddl_create_index.md#include-columns).
+
+## Connection pooling
+YugabyteDB uses the upper half of Postgresql to implement it's YSQL layer.
+Connection in Postgresql fork a new process, have at least 10MB of overhead in RAM, 
+have context-switching overhead and are generally slow to create. This is
+usually mitigated by keeping connection poolers in the client code or in a 
+ [proxy/loadbalancer](#ysql-proxy--load-balancer).
+
+Because of each connections overhead, we have to be careful regarding the number
+of opened connections that we keep for each tserver. Currently the maximum number 
+per-tserver is 300 and is set by the `--ysql_max_connections` gflag.
+
+We're also working to lower the overhead of connections.
+
```

Comment:
Other things that we should add to the doc include:

- YSQL requires PK to be specified in CREATE TABLE statement upfront.
- Use multi row inserts (i.e. inserts with multiple values in the VALUES clause) wherever possible.
- It is better to create indexes and foreign keys at the time of table creation instead of adding them after loading data.
- If your database has a large number of tables, then use colocation or reduce number of tablets per table using ysql_num_shards_per_tserver gflag.
- Use jsonb to group together less frequently accessed columns into a single column.
- Use UUIDs instead of auto-increment serial IDs.
- Use covered indexes where applicable to be able to do index only scans that are more performant.

Reply:
Added all above.